l10n_it_edi_doi_extension/models/purchase_order.py

```diff
@@ -0,0 +1,310 @@
+                order._l10n_it_edi_doi_get_amount_not_yet_invoiced(declaration)
+            )
+
+    @api.depends(
+        "l10n_it_edi_doi_id", "l10n_it_edi_doi_id.remaining", "state", "tax_totals"
+    )
+    def _compute_l10n_it_edi_doi_warning(self):
+        for order in self:
+            order.l10n_it_edi_doi_warning = ""
+            declaration = order.l10n_it_edi_doi_id
+
+            show_warning = declaration and order.state != "cancelled"
+            if not show_warning:
+                continue
+
+            declaration_not_yet_invoiced = declaration.not_yet_invoiced
+            if order.state == "purchase":
+                old_order_state = order._origin
+                declaration_not_yet_invoiced -= (
+                    old_order_state.l10n_it_edi_doi_not_yet_invoiced
+                )
+            declaration_not_yet_invoiced += order.l10n_it_edi_doi_not_yet_invoiced
+
+            validity_warnings = declaration._get_validity_warnings(
+                order.company_id,
+                order.partner_id.commercial_partner_id,
+                order.currency_id,
+                order.l10n_it_edi_doi_date,
+                sales_order=True,
+            )
+
+            threshold_warning = declaration._build_threshold_warning_message(
+                declaration.invoiced, declaration_not_yet_invoiced
+            )
+
+            order.l10n_it_edi_doi_warning = "{}\n\n{}".format(
+                "\n".join(validity_warnings), threshold_warning
+            ).strip()
+
+    @api.depends("l10n_it_edi_doi_id")
+    def _compute_fiscal_position_id(self):
+        super()._compute_fiscal_position_id()
+        for order in self:
+            declaration_fiscal_position = (
+                order.company_id.l10n_it_edi_doi_fiscal_position_id
+            )
+            if declaration_fiscal_position and order.l10n_it_edi_doi_id:
+                order.fiscal_position_id = declaration_fiscal_position
+        return  # W8110
+
+    def _prepare_invoice(self):
+        """
+        Prepare the dict of values to create the new invoice for a purchase order.
+        This method may be overridden to implement custom invoice generation
+        (making sure to call super() to establish a clean extension chain).
+        """
+        vals = super()._prepare_invoice()
+        declaration = self.l10n_it_edi_doi_id
+        if declaration:
+            date = fields.Date.context_today(self)
+            validity_warnings = declaration._get_validity_warnings(
+                self.company_id,
+                self.partner_id.commercial_partner_id,
+                self.currency_id,
+                date,
+                sales_order=True,
+            )
+            if not validity_warnings:
+                vals["l10n_it_edi_doi_id"] = declaration.id
+        return vals
+
+    def copy_data(self, default=None):
+        data_list = super().copy_data(default)
+        for order, data in zip(self, data_list, strict=False):
```

Comment:
```suggestion
        for order, data in zip(self, data_list):
```
~~Senza `strict` si rischia che `order` o `data` siano vuoti, come mai serve rimuoverlo?~~ vedi https://github.com/OCA/l10n-italy/pull/4680#discussion_r2081225525.

Reply:
?

Reply:
?

Reply:
@monen17 quel campo viene ri-aggiunto dal pre-commit. Ignoro quella riga specifica?

Reply:
Non sapevo venisse aggiunto da `pre-commit`, ho indagato e viene aggiunto per https://docs.astral.sh/ruff/rules/zip-without-explicit-strict, dove viene spiegato:
> By default, if the iterables passed to `zip` are of different lengths, the resulting iterator will be silently truncated to the length of the shortest iterable. This can lead to subtle bugs.

Quindi non è vero quello che scrissi in https://github.com/OCA/l10n-italy/pull/4680#discussion_r2045527104:
> Senza `strict` si rischia che `order` o `data` siano vuoti

Sapendo questo, si potrebbe mettere `strict=True` ma secondo me ci possiamo fidare che siano sempre della stessa lunghezza quindi va bene così :+1: